modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/DataStreamLifecycleService.java

```diff
@@ -310,7 +330,168 @@ void run(ClusterState state) {
                     e
                 );
             }
+
+            try {
+                maybeExecuteDownsampling(state, dataStream, getTargetIndices(dataStream, indicesToExcludeForRemainingRun));
+            } catch (Exception e) {
+                logger.error(
+                    () -> String.format(
+                        Locale.ROOT,
+                        "Data stream lifecycle failed to execute downsampling for data stream [%s]",
+                        dataStream.getName()
+                    ),
+                    e
+                );
+            }
+        }
+    }
+
+    Set<Index> maybeExecuteDownsampling(ClusterState state, DataStream dataStream, List<Index> targetIndices) {
+        Set<Index> affectedIndices = new HashSet<>();
+        Metadata metadata = state.metadata();
+        for (Index index : targetIndices) {
+            if (dataStream.isIndexManagedByDataStreamLifecycle(index, metadata::index)) {
+                IndexMetadata backingIndex = metadata.index(index);
+                assert backingIndex != null : "the data stream backing indices must exist";
+                List<DataStreamLifecycle.Downsampling.Round> downsamplingRounds = dataStream.getDownsamplingRoundsFor(
+                    index,
+                    metadata::index,
+                    nowSupplier
+                );
+                if (downsamplingRounds.isEmpty()) {
+                    continue;
+                }
+
+                String indexName = index.getName();
+                IndexMetadata.DownsampleTaskStatus backingIndexDownsamplingStatus = INDEX_DOWNSAMPLE_STATUS.get(backingIndex.getSettings());
+                String backingIndexDownsamplingSource = backingIndex.getSettings().get(IndexMetadata.INDEX_DOWNSAMPLE_SOURCE_NAME_KEY);
```

Comment:
Why do we use `.get(Setting)` here instead of the standardized `Setting.get(…)` methods? Like:

```suggestion
                String backingIndexDownsamplingSource = IndexMetadata.INDEX_DOWNSAMPLE_SOURCE_NAME.get(backingIndex.getSettings());
```

Reply:
Ah that setting doesn't have a default value, but should've used the proper way nonetheless. Great catch, thank you!